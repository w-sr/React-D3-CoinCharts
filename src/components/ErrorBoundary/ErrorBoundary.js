import React, { Component } from "react";
import PropTypes from "prop-types";
import Raven from "raven-js";
import styled from "styled-components";

import Flex from "../Flex";
import GithubBanner from "../GithubBanner";
import { color, fontFamily, fontSize, fontWeight } from "../../styles/constants";

const StyledErrorBoundary = styled(Flex)`
  font-family: ${fontFamily.regular};
  color: ${color.white};
  font-size: ${fontSize.huge};
  font-weight: ${fontWeight.medium};
  width: 100vw;
`;

const ERROR_MESSAGE_TEXT = "Uh-oh, Something went wrong...";

class ErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  componentDidCatch(error, info) {
    this.setState({ hasError: true });
    Raven.captureException(error, { extra: info });
  }

  render() {
    return this.state.hasError ? (
      <StyledErrorBoundary center>
        <GithubBanner />
        <div center>{ERROR_MESSAGE_TEXT}</div>
      </StyledErrorBoundary>
    ) : (
      this.props.children
    );
  }
}

ErrorBoundary.propTypes = {
  children: PropTypes.node.isRequired,
};

export default ErrorBoundary;
