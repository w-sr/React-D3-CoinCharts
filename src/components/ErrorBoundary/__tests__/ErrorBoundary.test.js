import React from "react";
import { render } from "react-testing-library";

import ErrorBoundary from "../";

describe("<ErrorBoundary />", () => {
  it("renders without crashing", () => {
    const { container } = render(
      <ErrorBoundary>
        <div>Hello World</div>
      </ErrorBoundary>,
    );

    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("throws console.error if there's no `props.children`", () => {
    console.error = jest.fn(); // eslint-disable-line no-console

    try {
      render(<ErrorBoundary />);
      throw new Error("should have thrown error");
    } catch (err) {
      expect(console.error).toHaveBeenCalled(); // eslint-disable-line no-console
    }
  });
});
