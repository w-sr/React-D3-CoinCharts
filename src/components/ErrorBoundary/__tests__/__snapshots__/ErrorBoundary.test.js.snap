// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorBoundary /> renders without crashing 1`] = `
<div>
  Hello World
</div>
`;
